File: Data-Structure/Tree/Heap/c/heap.c
#include "heap.h"

Heap* createHeap(int initialSize) {
    Heap* newHeap = (Heap*)malloc(sizeof(Heap));

    (*newHeap).nodes = (Node*)malloc(sizeof(Node)* initialSize);
    (*newHeap).capacity = initialSize;
    (*newHeap).usedSize = 0;

    printf("Size of Heap's Nodes %d \n", sizeof((*newHeap).nodes));

    return newHeap;
}
void destroyHeap(Heap* heap) {
    if (heap == NULL) return;

    free((*heap).nodes);
    free(heap);
}
void insert(Heap* heap, NodeData newData) {
    int currentPosition = (*heap).usedSize;
    int parentPosition = getParent(currentPosition);

    if ( currentPosition == (*heap).capacity ) {

        (*heap).capacity *= 2;
        (*heap).nodes = (Node*)realloc((*heap).nodes, (sizeof(Node) * (*heap).capacity));
    }

    (*(((*heap).nodes) + currentPosition)).data = newData;

    while ( currentPosition > 0 &&
        (*(((*heap).nodes) + currentPosition)).data < (*(((*heap).nodes) + parentPosition)).data 
    ) {
        NodeData currentsTemp = (*(((*heap).nodes) + currentPosition)).data;

        swapNodes(heap, currentPosition, parentPosition);

        currentPosition = parentPosition;
        parentPosition = getParent(currentPosition);

    }
}

void deleteMin(Heap* heap, Node* root) {
    int parentPosition = 0;
    int leftPosition = 0;
    int rightPosition = 0;

    memcpy(root, (*heap).nodes, sizeof(Node));
    memset((*heap).nodes, 0, sizeof(Node));

    (*heap).usedSize--;

    swapNodes(heap, 0, (*heap).usedSize);

    leftPosition = getLeftChild(0);
    rightPosition = leftPosition + 1;

    while (1) {
        int selectedChild = 0;

        if (leftPosition >= (*heap).usedSize) break;

        if (rightPosition >= (*heap).usedSize) {
         selectedChild = leftPosition;
        } else {

            if ( (*((*heap).nodes + leftPosition)).data > (*((*heap).nodes + rightPosition)).data ) {
                selectedChild = rightPosition;
            } else {
                selectedChild = leftPosition;
            }

        }

        if ( (*((*heap).nodes + parentPosition)).data > (*((*heap).nodes + selectedChild)).data ) {
            swapNodes(heap, parentPosition, selectedChild);
            parentPosition = selectedChild;
        }else {
            break;
        }

        leftPosition = getLeftChild(parentPosition);
        rightPosition = leftPosition + 1;
    }

    if ( (*heap).usedSize < (*heap).capacity / 2 ) {
            (*heap).capacity /= 2;

            (*heap).nodes = (Node*)realloc((*heap).nodes, sizeof(Node) * (*heap).capacity);
    }
}

int getParent(int index) {
    return (index -1) / 2;
}
int getLeftChild(int index) {
    return (index * 2) + 1;
}

void swapNodes(Heap* heap, int indexF, int indexS) {
    int copySize = sizeof(Node);
    Node* tempNode = (Node*)malloc(copySize);

    memcpy(
        tempNode, 
        ((*heap).nodes) + indexF,
        copySize
    );

    memcpy(
        ((*heap).nodes) + indexF,
        ((*heap).nodes) + indexS,
        copySize
    );

    memcpy(
        ((*heap).nodes) + indexS,
        tempNode,
        copySize
    );

    free(tempNode);
}

void printNodes(Heap* heap) {
    for (int i=0; i< (*heap).usedSize; i++) {
        printf("%d ", (*((*heap).nodes + i)).data );
    }
    puts("");
}

File: Data-Structure/Tree/Heap/rust-procedural/heap.rs
pub type nodeDT = i32;

pub struct Heap {
    nodes: Vec<nodeDT>,
}

pub fn create_heap(initial_size: usize) -> Heap {
    return Heap {
        nodes: Vec::with_capacity(initial_size),
    };
}

fn get_parent(index: usize) -> usize {
    if index == 0 {
        return 0;
    } else {
        return (index - 1) / 2;
    }
}

fn get_left_chlid(index: usize) -> usize {
    return (index * 2) + 1;
}

pub fn insert(heap: &mut Heap, new_data: nodeDT) {
    heap.nodes.push(new_data);

    let mut current_position = heap.nodes.len() - 1;
    let mut parent_position = get_parent(current_position);

    while current_position > 0 && heap.nodes[current_position] < heap.nodes[parent_position] {
        heap.nodes.swap(current_position, parent_position);
        current_position = parent_position;
        parent_position = get_parent(current_position);
    }
}

pub fn delete_min(heap: &mut Heap) -> Option<nodeDT> {
    let len = heap.nodes.len();

    if len == 0 {
        return None;
    }

    let min_data = heap.nodes[0];

    if len == 1 {
        heap.nodes.pop();
        return Some(min_data);
    }

    heap.nodes.swap_remove(0);


    let mut parent_position = 0;

    loop {
        let left_position = get_left_chlid(parent_position);

        if left_position >= heap.nodes.len() {
            break;
        }

        let right_position = left_position + 1;

        let mut selected_child = left_position;
        if right_position < heap.nodes.len() && heap.nodes[right_position] < heap.nodes[left_position] {
            selected_child = right_position;
        }

        if heap.nodes[parent_position] > heap.nodes[selected_child] {
            heap.nodes.swap(parent_position, selected_child);
            parent_position = selected_child;
        }else {
            break;
        }
    }

    return Some(min_data);
}

pub fn print_nodes(heap: &Heap) {
    for data in &heap.nodes {
        print!("{} ", data);
    }
    println!();
}
